pageserver/src/tenant/mgr.rs

```diff
@@ -2160,94 +2187,56 @@ pub(crate) async fn get_active_tenant_with_timeout(
     timeout: Duration,
     cancel: &CancellationToken,
 ) -> Result<Arc<Tenant>, GetActiveTenantError> {
-    enum WaitFor {
-        Barrier(utils::completion::Barrier),
-        Tenant(Arc<Tenant>),
-    }
-
     let wait_start = Instant::now();
     let deadline = wait_start + timeout;
 
-    let (wait_for, tenant_shard_id) = {
-        let locked = TENANTS.read().unwrap();
-
-        // Resolve TenantId to TenantShardId
-        let tenant_shard_id = locked
-            .resolve_attached_shard(&tenant_id, shard_selector)
-            .ok_or(GetActiveTenantError::NotFound(GetTenantError::NotFound(
-                tenant_id,
-            )))?;
-
-        let peek_slot = tenant_map_peek_slot(&locked, &tenant_shard_id, TenantSlotPeekMode::Read)
-            .map_err(GetTenantError::MapState)?;
-        match peek_slot {
-            Some(TenantSlot::Attached(tenant)) => {
-                match tenant.current_state() {
-                    TenantState::Active => {
-                        // Fast path: we don't need to do any async waiting.
-                        return Ok(tenant.clone());
-                    }
-                    _ => {
-                        tenant.activate_now();
-                        (WaitFor::Tenant(tenant.clone()), tenant_shard_id)
-                    }
-                }
-            }
-            Some(TenantSlot::Secondary(_)) => {
-                return Err(GetActiveTenantError::NotFound(GetTenantError::NotActive(
-                    tenant_shard_id,
-                )))
-            }
-            Some(TenantSlot::InProgress(barrier)) => {
-                (WaitFor::Barrier(barrier.clone()), tenant_shard_id)
-            }
-            None => {
+    // Resolve TenantId to TenantShardId.  This is usually a quick one-shot thing, the loop is
+    // for handling the rare case that the slot we're accessing is InProgress.
+    let tenant_shard = loop {
+        let resolved = {
+            let locked = TENANTS.read().unwrap();
+            locked.resolve_attached_shard(&tenant_id, shard_selector)
+        };
+        match resolved {
+            ShardResolveResult::Found(tsid) => break tsid,
```

Comment:
Weird variable name `tsid`, this is an object, not the id


```suggestion
            ShardResolveResult::Found(tenant_shard) => break tenant_shard,
```

Reply:
Ah, failed to update this when the function changed from returning an ID to returning a Tenant.   Applied by hand in 0d4247d2d8242bc2767cdebdeaa34a7d9a9ffb16